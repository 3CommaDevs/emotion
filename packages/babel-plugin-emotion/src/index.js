// @flow
import fs from 'fs'
import nodePath from 'path'
import { touchSync } from 'touch'
import { addSideEffect } from '@babel/helper-module-imports'
import {
  getIdentifierName,
  getName,
  createRawStringFromTemplateLiteral,
  minify
} from './babel-utils'
import type {
  BabelPath as _BabelPath,
  Identifier,
  BabelPluginPass,
  Types,
  StringLiteral,
  Babel
} from 'babel-flow-types'
import { hashString, Stylis } from 'emotion-utils'
import { addSourceMaps } from './source-map'

import cssProps from './css-prop'
import ASTObject from './ast-object'

export type BabelPath = _BabelPath & {
  node: *
}

export function hashArray(arr: Array<string>) {
  return hashString(arr.join(''))
}

const staticStylis = new Stylis({ keyframe: false })

export function hoistPureArgs(path: BabelPath) {
  const args = path.get('arguments')

  if (args && Array.isArray(args)) {
    args.forEach(arg => {
      if (!arg.isIdentifier() && arg.isPure()) {
        arg.hoist()
      }
    })
  }
}

type ImportedNames = {
  css: string,
  keyframes: string,
  injectGlobal: string,
  styled: string,
  cx: string
}

export type EmotionBabelPluginPass = BabelPluginPass & {
  extractStatic: boolean,
  insertStaticRules: (rules: Array<string>) => void,
  emotionImportPath: string,
  staticRules: Array<string>,
  cssPropIdentifier: Identifier,
  cssPropCxIdentifier: Identifier,
  importedNames: ImportedNames
}

export function replaceCssWithCallExpression(
  path: BabelPath,
  identifier: Identifier,
  state: EmotionBabelPluginPass,
  t: Types,
  staticCSSSrcCreator: (
    src: string,
    name: string,
    hash: string
  ) => string = src => src,
  removePath: boolean = false,
  staticCSSSelectorCreator: (name: string, hash: string) => string = (
    name,
    hash
  ) => `.${name}-${hash}`
) {
  try {
    let { hash, src } = createRawStringFromTemplateLiteral(path.node.quasi)
    const identifierName = getIdentifierName(path, t)
    const name = getName(identifierName, 'css')

    if (state.extractStatic && !path.node.quasi.expressions.length) {
      const staticCSSRules = staticStylis(
        staticCSSSelectorCreator(name, hash),
        staticCSSSrcCreator(src, name, hash)
      )
      state.insertStaticRules([staticCSSRules])
      if (!removePath) {
        return path.replaceWith(t.stringLiteral(`${name}-${hash}`))
      }
      return path.replaceWith(t.identifier('undefined'))
    }

    if (!removePath) {
      path.addComment('leading', '#__PURE__')
    }
    if (state.opts.sourceMap === true && path.node.quasi.loc !== undefined) {
      src += addSourceMaps(path.node.quasi.loc.start, state)
    }

    path.replaceWith(
      t.callExpression(
        identifier,
        new ASTObject(minify(src), path.node.quasi.expressions, t)
          .toExpressions()
          .concat(
            state.opts.autoLabel && identifierName
              ? [t.stringLiteral(`label:${identifierName.trim()};`)]
              : []
          )
      )
    )

    if (state.opts.hoist) {
      hoistPureArgs(path)
    }

    return
  } catch (e) {
    if (path) {
      throw path.buildCodeFrameError(e)
    }

    throw e
  }
}

export function buildStyledCallExpression(
  identifier: Identifier,
  tag: StringLiteral,
  path: BabelPath,
  state: EmotionBabelPluginPass,
  t: Types
) {
  const identifierName = getIdentifierName(path, t)

  if (state.extractStatic && !path.node.quasi.expressions.length) {
    const { hash, src } = createRawStringFromTemplateLiteral(path.node.quasi)
    const staticClassName = `css-${hash}`
    const staticCSSRules = staticStylis(`.${staticClassName}`, src)

    state.insertStaticRules([staticCSSRules])
    return t.callExpression(
      t.callExpression(identifier, [
        tag,
        t.objectExpression([
          t.objectProperty(t.identifier('e'), t.stringLiteral(staticClassName))
        ])
      ]),
      []
    )
  }

  let { src } = createRawStringFromTemplateLiteral(path.node.quasi)

  path.addComment('leading', '#__PURE__')

  if (state.opts.sourceMap === true && path.node.quasi.loc !== undefined) {
    src += addSourceMaps(path.node.quasi.loc.start, state)
  }

  return t.callExpression(
    t.callExpression(
      identifier,
      state.opts.autoLabel && identifierName
        ? [
            tag,
            t.objectExpression([
              t.objectProperty(
                t.identifier('label'),
                t.stringLiteral(identifierName.trim())
              )
            ])
          ]
        : [tag]
    ),
    new ASTObject(minify(src), path.node.quasi.expressions, t).toExpressions()
  )
}

export function buildStyledObjectCallExpression(
  path: BabelPath,
  state: EmotionBabelPluginPass,
  identifier: Identifier,
  t: Types
) {
  const identifierName = getIdentifierName(path, t)
  const tag = t.isCallExpression(path.node.callee)
    ? path.node.callee.arguments[0]
    : t.stringLiteral(path.node.callee.property.name)

  let args = path.node.arguments
  if (state.opts.sourceMap === true && path.node.loc !== undefined) {
    args.push(t.stringLiteral(addSourceMaps(path.node.loc.start, state)))
  }

  path.addComment('leading', '#__PURE__')

  return t.callExpression(
    t.callExpression(
      identifier,
      state.opts.autoLabel && identifierName
        ? [
            tag,
            t.objectExpression([
              t.objectProperty(
                t.identifier('label'),
                t.stringLiteral(identifierName.trim())
              )
            ])
          ]
        : [tag]
    ),
    args
  )
}

const visited = Symbol('visited')

const defaultImportedNames: ImportedNames = {
  styled: 'styled',
  css: 'css',
  keyframes: 'keyframes',
  injectGlobal: 'injectGlobal',
  cx: 'cx'
}

const importedNameKeys = Object.keys(defaultImportedNames)

const defaultEmotionPaths = ['emotion', 'react-emotion', 'preact-emotion']

function getPath(filepath: string) {
  if (filepath.charAt(0) === '.') {
    return nodePath.resolve(process.cwd(), filepath)
  }
  return filepath
}

export default function(babel: Babel) {
  const { types: t } = babel

  return {
    name: 'emotion', // not required
    inherits: require('babel-plugin-syntax-jsx'),
    visitor: {
      Program: {
        enter(path: BabelPath, state: EmotionBabelPluginPass) {
          // this needs to handle relative paths and stuff
          // https://github.com/tleunen/babel-plugin-module-resolver/tree/master/src
          state.emotionImportPath = 'emotion'
          if (state.opts.primaryPath !== undefined) {
            state.emotionImportPath = getPath(state.opts.primaryPath)
          }

          state.importedNames = {
            ...defaultImportedNames,
            ...state.opts.importedNames
          }

          const imports = []

          let isModule = false

          for (const node of path.node.body) {
            if (t.isModuleDeclaration(node)) {
              isModule = true
              break
            }
          }

          if (isModule) {
            path.traverse({
              ImportDeclaration: {
                exit(path) {
                  const { node } = path

                  const imported = []
                  const specifiers = []

                  imports.push({
                    source: node.source.value,
                    imported,
                    specifiers
                  })

                  for (const specifier of path.get('specifiers')) {
                    const local = specifier.node.local.name

                    if (specifier.isImportDefaultSpecifier()) {
                      imported.push('default')
                      specifiers.push({
                        kind: 'named',
                        imported: 'default',
                        local
                      })
                    }

                    if (specifier.isImportSpecifier()) {
                      const importedName = specifier.node.imported.name
                      imported.push(importedName)
                      specifiers.push({
                        kind: 'named',
                        imported: importedName,
                        local
                      })
                    }
                  }
                }
              }
            })
          }
          const emotionPaths = defaultEmotionPaths.concat(
            (state.opts.paths || []).map(getPath)
          )
          imports.forEach(({ source, imported, specifiers }) => {
            if (emotionPaths.indexOf(getPath(source)) !== -1) {
              const importedNames = specifiers
                .filter(v => importedNameKeys.indexOf(v.imported) !== -1)
                .reduce(
                  (acc, { imported, local }) => ({
                    ...acc,
                    [imported === 'default' ? 'styled' : imported]: local
                  }),
                  defaultImportedNames
                )
              state.importedNames = {
                ...importedNames,
                ...state.opts.importedNames
              }
            }
          })

          state.extractStatic =
            // path.hub.file.opts.filename !== 'unknown' ||
            state.opts.extractStatic

          state.staticRules = []

          state.insertStaticRules = function(staticRules) {
            state.staticRules.push(...staticRules)
          }
        },
        exit(path: BabelPath, state: EmotionBabelPluginPass) {
          if (state.staticRules.length !== 0) {
            const toWrite = state.staticRules.join('\n').trim()
            const filenameArr = path.hub.file.opts.filename.split('.')
            filenameArr.pop()
            filenameArr.push('emotion', 'css')
            const cssFilename = filenameArr.join('.')
            const exists = fs.existsSync(cssFilename)
            addSideEffect(path, './' + nodePath.basename(cssFilename))
            if (
              exists ? fs.readFileSync(cssFilename, 'utf8') !== toWrite : true
            ) {
              if (!exists) {
                touchSync(cssFilename)
              }
              fs.writeFileSync(cssFilename, toWrite)
            }
          }
        }
      },
      JSXOpeningElement(path: BabelPath, state: EmotionBabelPluginPass) {
        cssProps(path, state, t)
        if (state.opts.hoist) {
          path.traverse({
            CallExpression(callExprPath) {
              if (
                callExprPath.node.callee.name === state.importedNames.css ||
                callExprPath.node.callee === state.cssPropIdentifier
              ) {
                hoistPureArgs(callExprPath)
              }
            }
          })
        }
      },
      CallExpression(path: BabelPath, state: EmotionBabelPluginPass) {
        // $FlowFixMe
        if (path[visited]) {
          return
        }
        // $FlowFixMe
        path[visited] = true

        try {
          if (t.isIdentifier(path.node.callee)) {
            switch (path.node.callee.name) {
              case state.importedNames.css:
              case state.importedNames.keyframes: {
                path.addComment('leading', '#__PURE__')
                if (state.opts.autoLabel) {
                  const identifierName = getIdentifierName(path, t)
                  if (identifierName) {
                    path.node.arguments.push(
                      t.stringLiteral(`label:${identifierName.trim()};`)
                    )
                  }
                }
              }
              // eslint-disable-next-line no-fallthrough
              case state.importedNames.injectGlobal:
                if (
                  state.opts.sourceMap === true &&
                  path.node.loc !== undefined
                ) {
                  path.node.arguments.push(
                    t.stringLiteral(addSourceMaps(path.node.loc.start, state))
                  )
                }
            }
          }

          if (
            (t.isCallExpression(path.node.callee) &&
              path.node.callee.callee.name === state.importedNames.styled) ||
            (t.isMemberExpression(path.node.callee) &&
              t.isIdentifier(path.node.callee.object) &&
              path.node.callee.object.name === state.importedNames.styled)
          ) {
            const identifier = t.isCallExpression(path.node.callee)
              ? path.node.callee.callee
              : path.node.callee.object
            path.replaceWith(
              buildStyledObjectCallExpression(path, state, identifier, t)
            )

            if (state.opts.hoist) {
              hoistPureArgs(path)
            }
          }
        } catch (e) {
          throw path.buildCodeFrameError(e)
        }
      },
      TaggedTemplateExpression(path: BabelPath, state: EmotionBabelPluginPass) {
        // $FlowFixMe
        if (path[visited]) {
          return
        }
        // $FlowFixMe
        path[visited] = true
        if (
          // styled.h1`color:${color};`
          t.isMemberExpression(path.node.tag) &&
          path.node.tag.object.name === state.importedNames.styled
        ) {
          path.replaceWith(
            buildStyledCallExpression(
              path.node.tag.object,
              t.stringLiteral(path.node.tag.property.name),
              path,
              state,
              t
            )
          )
        } else if (
          // styled('h1')`color:${color};`
          t.isCallExpression(path.node.tag) &&
          path.node.tag.callee.name === state.importedNames.styled
        ) {
          path.replaceWith(
            buildStyledCallExpression(
              path.node.tag.callee,
              path.node.tag.arguments[0],
              path,
              state,
              t
            )
          )
        } else if (t.isIdentifier(path.node.tag)) {
          if (
            path.node.tag.name === state.importedNames.css ||
            path.node.tag === state.cssPropIdentifier
          ) {
            replaceCssWithCallExpression(path, path.node.tag, state, t)
          } else if (path.node.tag.name === state.importedNames.keyframes) {
            replaceCssWithCallExpression(
              path,
              path.node.tag,
              state,
              t,
              (src, name, hash) => `@keyframes ${name}-${hash} { ${src} }`,
              false,
              () => ''
            )
          } else if (path.node.tag.name === state.importedNames.injectGlobal) {
            replaceCssWithCallExpression(
              path,
              path.node.tag,
              state,
              t,
              undefined,
              true,
              () => ''
            )
          }
        }
      }
    }
  }
}
