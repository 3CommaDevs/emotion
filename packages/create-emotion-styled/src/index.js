import { memoize } from 'emotion-utils'
import type { Emotion, Interpolation, Interpolations } from 'create-emotion'

function setTheme(theme) {
  this.setState({ theme })
}

declare var codegen: { require: (path: string) => * }

const reactPropsRegex: RegExp = codegen.require('./props')
const testOmitPropsOnStringTag: (key: string) => boolean = memoize(key =>
  reactPropsRegex.test(key)
)
const testOmitPropsOnComponent = key => key !== 'theme' && key !== 'innerRef'
const testAlwaysTrue = () => true

const omitAssign: (
  testFn: (key: string) => boolean,
  target: {},
  ...sources: Array<{}>
) => Object = function(testFn, target) {
  let i = 2
  let length = arguments.length
  for (; i < length; i++) {
    let source = arguments[i]
    let key
    for (key in source) {
      if (testFn(key)) {
        target[key] = source[key]
      }
    }
  }
  return target
}

export type EmotionStyledOptions = {
  channel: string,
  contextTypes?: *,
  Component: *,
  createElement: Function,
  contextTypes?: *
}

function createEmotionStyled(
  emotion: Emotion,
  instanceOptions: EmotionStyledOptions
) {
  function componentWillMount() {
    if (this.context[instanceOptions.channel] !== undefined) {
      this.unsubscribe = this.context[instanceOptions.channel].subscribe(
        setTheme.bind(this)
      )
    }
  }
  function componentWillUnmount() {
    if (this.unsubscribe !== undefined) {
      this.context[instanceOptions.channel].unsubscribe(this.unsubscribe)
    }
  }
  const createStyled = (tag: *, options: { e: string, label: string }) => {
    if (process.env.NODE_ENV !== 'production') {
      if (tag === undefined) {
        throw new Error(
          'You are trying to create a styled element with an undefined component.\nYou may have forgotten to import it.'
        )
      }
    }
    let staticClassName
    let identifierName
    if (options !== undefined) {
      staticClassName = options.e
      identifierName = options.label
    }
    const isReal = tag.__emotion_real === tag
    const baseTag =
      staticClassName === undefined
        ? (isReal && tag.__emotion_base) || tag
        : tag

    const omitFn =
      typeof baseTag === 'string' &&
      baseTag.charAt(0) === baseTag.charAt(0).toLowerCase()
        ? testOmitPropsOnStringTag
        : testOmitPropsOnComponent

    return (strings: Interpolation, ...interpolations: Interpolations) => {
      let styles = (isReal && tag.__emotion_styles) || []
      if (identifierName !== undefined) {
        styles = styles.concat(`label:${identifierName};`)
      }
      if (staticClassName === undefined) {
        if (strings == null || strings.raw === undefined) {
          styles = styles.concat(strings, interpolations)
        } else {
          styles = interpolations.reduce(
            (array, interp, i) => array.concat(interp, strings[i + 1]),
            styles.concat(strings[0])
          )
        }
      }

      class Styled extends instanceOptions.Component {
        render() {
          const { props, state } = this
          this.mergedProps = omitAssign(testAlwaysTrue, {}, props, {
            theme: (state !== null && state.theme) || props.theme || {}
          })

          let className = ''
          let classInterpolations = []

          if (props.className) {
            if (staticClassName === undefined) {
              className += emotion.getRegisteredStyles(
                classInterpolations,
                props.className
              )
            } else {
              className += `${props.className} `
            }
          }
          if (staticClassName === undefined) {
            className += emotion.css.apply(
              this,
              styles.concat(classInterpolations)
            )
          } else {
            className += staticClassName
          }

          return instanceOptions.createElement(
            baseTag,
            omitAssign(omitFn, {}, props, { className, ref: props.innerRef })
          )
        }
      }
      Styled.prototype.componentWillMount = componentWillMount
      Styled.prototype.componentWillUnmount = componentWillUnmount
      Styled.displayName =
        identifierName !== undefined
          ? identifierName
          : `Styled(${typeof baseTag === 'string'
              ? baseTag
              : baseTag.displayName || baseTag.name || 'Component'})`

      Styled.contextTypes = instanceOptions.contextTypes
      Styled.__emotion_styles = styles
      Styled.__emotion_base = baseTag
      Styled.__emotion_real = Styled

      Styled.withComponent = nextTag => {
        return createStyled(nextTag, options)(styles)
      }

      return Styled
    }
  }
  return createStyled
}

export default createEmotionStyled
