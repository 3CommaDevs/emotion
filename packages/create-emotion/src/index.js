// @flow
import { hashString, Stylis } from 'emotion-utils'
import stylisRuleSheet from 'stylis-rule-sheet'
import { processStyleName, processStyleValue, classnames } from './utils'
import StyleSheet from './sheet'

type StylisPlugins = Function[]

type EmotionCaches = {
  registered: { [key: string]: string },
  inserted: { [key: string]: string | true },
  stylis: (scope: string, styles: string) => string,
  sheet: StyleSheet,
  externalStylisPlugins: StylisPlugins
}

function createEmotion(
  context: { __SECRET_EMOTION__: EmotionCaches },
  options?: { nonce?: string }
) {
  if (options === undefined) options = {}
  let caches: EmotionCaches = context.__SECRET_EMOTION__
  let current

  function insertRule(rule: string) {
    current += rule
    sheet.insert(rule, currentSourceMap)
  }

  const insertionPlugin = stylisRuleSheet(insertRule)

  if (caches === undefined) {
    const stylisOptions: { keyframe?: boolean, compress?: boolean } = {
      keyframe: false
    }

    if (process.env.NODE_ENV !== 'production') {
      stylisOptions.compress = false
    }
    context.__SECRET_EMOTION__ = caches = {
      registered: {},
      inserted: {},
      sheet: new StyleSheet(options.nonce),
      externalStylisPlugins: [],
      stylis: new Stylis(stylisOptions)
    }

    caches.stylis.use(insertionPlugin)
    // ðŸš€
    caches.sheet.inject()
  }

  const useStylisPlugin = (plugin: Function) => {
    caches.externalStylisPlugins.push(plugin)
    caches.stylis.use(null)(caches.externalStylisPlugins)(insertionPlugin)
  }

  let stylis = caches.stylis
  let sheet = caches.sheet

  let registered = {}

  let inserted = {}

  let currentSourceMap = ''

  type Interpolation =
    | string
    | number
    | Object
    | Interpolation[]
    | (() => Interpolation)
    | ((props?: Object, context?: Object) => Interpolation)
  function handleInterpolation(
    interpolation: Interpolation,
    couldBeSelectorInterpolation: boolean
  ): string | number {
    if (interpolation == null) {
      return ''
    }

    switch (typeof interpolation) {
      case 'boolean':
        return ''
      case 'function':
        return handleInterpolation.call(
          this,
          this === undefined
            ? interpolation()
            : // $FlowFixMe
              interpolation(this.mergedProps, this.context),
          couldBeSelectorInterpolation
        )
      case 'object':
        return createStringFromObject.call(this, interpolation)
      default:
        const cached: string | void = registered[interpolation]
        return couldBeSelectorInterpolation === false && cached !== undefined
          ? cached
          : interpolation
    }
  }

  const objectToStringCache = new WeakMap()

  function createStringFromObject(obj: Object | any[]): string {
    if (objectToStringCache.has(obj)) {
      // $FlowFixMe
      return objectToStringCache.get(obj)
    }
    let string = ''

    if (Array.isArray(obj)) {
      obj.forEach(function(interpolation: Interpolation) {
        string += handleInterpolation.call(this, interpolation, false)
      }, this)
    } else {
      Object.keys(obj).forEach(function(key: string) {
        // $FlowFixMe
        if (typeof obj[key] !== 'object') {
          if (registered[obj[key]] !== undefined) {
            string += `${key}{${registered[obj[key]]}}`
          } else {
            string += `${processStyleName(key)}:${processStyleValue(
              key,
              // $FlowFixMe
              obj[key]
            )};`
          }
        } else {
          string += `${key}{${handleInterpolation.call(this, obj[key], false)}}`
        }
      }, this)
    }
    objectToStringCache.set(obj, string)

    return string
  }

  let name

  const labelPattern = /label:\s*([^\s;\n]+)\s*[;\n]/g

  function createStyles(
    strings: Interpolation | string[],
    ...interpolations: Interpolation[]
  ) {
    let stringMode = true
    let styles: string = ''
    let identifierName = ''

    if (strings == null || strings.raw === undefined) {
      stringMode = false
      // $FlowFixMe
      styles += handleInterpolation.call(this, strings, false)
    } else {
      // $FlowFixMe
      styles += strings[0]
    }

    interpolations.forEach(function(interpolation, i) {
      styles += handleInterpolation.call(
        this,
        interpolation,
        styles.charCodeAt(styles.length - 1) === 46 // .
      )
      // $FlowFixMe
      if (stringMode === true && strings[i + 1] !== undefined) {
        // $FlowFixMe
        styles += strings[i + 1]
      }
    }, this)
    styles = styles.replace(labelPattern, (match, p1: string) => {
      identifierName += `-${p1}`
      return ''
    })
    name = hashString(styles + identifierName) + identifierName
    return styles
  }

  if (process.env.NODE_ENV !== 'production') {
    const sourceMapRegEx = /\/\*#\ssourceMappingURL=data:application\/json;\S+\s+\*\//
    const oldStylis = stylis
    stylis = (selector, styles) => {
      const result = sourceMapRegEx.exec(styles)
      currentSourceMap = result ? result[0] : ''
      oldStylis(selector, styles)
      currentSourceMap = ''
    }
  }
  function insert(scope, styles) {
    if (inserted[name] === undefined) {
      current = ''
      stylis(scope, styles)
      inserted[name] = current
    }
  }
  function css(...args: any) {
    const styles = createStyles.apply(this, arguments)
    const selector = `css-${name}`

    if (registered[selector] === undefined) {
      registered[selector] = styles
    }
    insert(`.${selector}`, styles)

    return selector
  }

  function keyframes() {
    const styles = createStyles.apply(this, arguments)
    const animation = `animation-${name}`
    insert('', `@keyframes ${animation}{${styles}}`)

    return animation
  }

  function injectGlobal() {
    const styles = createStyles.apply(this, arguments)
    insert('', styles)
  }

  function fontFace(...args: any) {
    const styles = createStyles(...args)
    insert('', `@font-face{${styles}}`)
  }

  function getRegisteredStyles(registeredStyles: string[], classNames: string) {
    let rawClassName = ''

    classNames.split(' ').forEach(className => {
      if (registered[className] !== undefined) {
        registeredStyles.push(className)
      } else {
        rawClassName += `${className} `
      }
    })
    return rawClassName
  }

  function merge(className: string, sourceMap?: string) {
    const registeredStyles = []

    const rawClassName = getRegisteredStyles(registeredStyles, className)

    if (registeredStyles.length < 2) {
      return className
    }
    return rawClassName + css(registeredStyles, sourceMap)
  }

  function cx(...classNames: any) {
    return merge(classnames(...classNames))
  }

  function hydrate(ids: string[]) {
    ids.forEach(id => {
      inserted[id] = true
    })
  }

  function flush() {
    sheet.flush()
    inserted = context.__SECRET_EMOTION__.inserted = {}
    registered = context.__SECRET_EMOTION__.registered = {}
    sheet.inject()
  }
  return {
    flush,
    hydrate,
    cx,
    merge,
    getRegisteredStyles,
    fontFace,
    injectGlobal,
    keyframes,
    css,
    sheet,
    registered,
    inserted,
    useStylisPlugin
  }
}

export default createEmotion
