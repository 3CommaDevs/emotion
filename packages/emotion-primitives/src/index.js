import reactPrimitives from 'react-primitives'

import { createEmotionPrimitive } from './createEmotion'
import { splitProps } from './splitProps'

let validate

const primitives = ['Text', 'View', 'Image']

const assignPrimitives = styled => {
  Object.assign(
    styled,
    primitives.reduce((getters, alias) => {
      getters[alias] = styled(reactPrimitives[alias])
      return getters
    }, {})
  )

  return styled
}

const styled = createEmotionPrimitive(splitProps)

if (process.env.NODE_ENV !== 'production' && typeof Proxy !== 'undefined') {
  const Platform = reactPrimitives.Platform

  // Validate primitives accessed using the emotion function directly like emotion.TEXT`` or emotion.VIEW``
  // Proxy is not supported in Native or Sketch. So makes sure that errors are not logged out in those env.
  if (Platform.OS === 'web') {
    validate = target => {
      const handler = {
        get: (obj, prop) => {
          if (prop in obj) {
            return obj[prop]
          } else {
            throw new Error(
              `Cannot style invalid primitive ${prop}. Expected primitive to be one of ['Text', 'View', 'Image']`
            )
          }
        }
      }

      return new Proxy(target, handler)
    }
  } else {
    validate = fn => fn
  }
}

export default validate(assignPrimitives(styled))
