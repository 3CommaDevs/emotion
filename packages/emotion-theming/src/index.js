// @flow
export { createThemeProvider } from './theme-provider'
export { createWithTheme } from './with-theme'
export { channel, contextTypes } from './utils'
