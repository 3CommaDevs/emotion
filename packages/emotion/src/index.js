import createEmotion from 'create-emotion'

const context =
  typeof global !== 'undefined'
    ? global
    : typeof window !== 'undefined' ? window : {}

export const {
  flush,
  hydrate,
  cx,
  merge,
  getRegisteredStyles,
  fontFace,
  injectGlobal,
  keyframes,
  css,
  sheet,
  registered,
  inserted,
  useStylisPlugin
} = createEmotion(context)
