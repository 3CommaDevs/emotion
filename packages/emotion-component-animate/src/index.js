import { createElement as h, Component } from 'react'
import { css, keyframes, merge } from 'emotion'

export default class Animate extends Component {
  constructor(props) {
    super(props)

    this.state = {
      cls: css()
    }

    this.styleEntries = Object.entries(props.styles)
    this.prevStyle =
      typeof props.style === 'function'
        ? props.style(this.props, this.context)
        : props.style
  }

  componentWillMount() {
    this.updateStyles(this.props)
  }

  componentWillReceiveProps(nextProps) {
    this.updateStyles(nextProps)
  }

  updateStyles = props => {
    const nextStyleEntries = Object.entries(props.style)
    if (
      nextStyleEntries.some(
        ([key, value], i) =>
          this.styleEntries[i][0] !== key || this.styleEntries[i][1] !== value
      )
    ) {
      const { style: styleProp, duration, timing } = props
      const style =
        typeof styleProp === 'function'
          ? styleProp(this.props, this.context)
          : styleProp

      const name = keyframes({
        from: this.prevStyle,
        to: style
      })

      this.setState({
        cls: css(
          { animationName: name },
          duration && { animationDuration: duration },
          timing && { animationTimingFunction: timing }
        )
      })

      this.prevStyle = style
      this.styleEntries = nextStyleEntries
    }
  }

  render() {
    const { cls } = this.state
    const { className, tag } = this.props
    return h(
      tag,
      Object.assign({}, this.props, {
        className: merge([className, cls].filter(Boolean).join(' '))
      })
    )
  }
}
