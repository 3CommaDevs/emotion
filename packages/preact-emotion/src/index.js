import _preact from 'preact'
import * as emotion from 'emotion'
import createEmotionStyled from 'create-emotion-styled'

const preact = {
  ..._preact,
  Children: {
    only: children => children[0]
  }
}

export default createEmotionStyled(emotion, preact)

export * from 'emotion'
