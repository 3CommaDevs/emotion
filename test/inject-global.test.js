import { stringify, parse } from 'css'
import { injectGlobal, sheet, flush, css } from '../src/index'

const getCss = () =>
  stringify(parse(sheet.tags.map(tag => tag.textContent || '').join('')))

describe('injectGlobal', () => {
  afterEach(() => {
    flush()
  })
  test('static', () => {
    injectGlobal`
      html {
        background: pink;
      }
      html.active {
        background: red;
      }
    `
    expect(getCss()).toMatchSnapshot()
  })
  test('interpolation', () => {
    const color = 'yellow'
    injectGlobal`
      body {
        color: ${color};
        margin: 0;
        padding: 0;
      }
    `
    expect(getCss()).toMatchSnapshot()
  })
  test('composition', () => {
    const style = css`
      background-color: hotpink;
      &:hover {
        background-color: palevioletred;
      }
    `
    injectGlobal`
      html {
        background: pink;
      }
      html.active {
        color: green;
        ${style};
      }
    `
    console.log(sheet.tags.map(tag => tag.textContent || '').join(''))
    expect(getCss()).toMatchSnapshot()
  })
})
