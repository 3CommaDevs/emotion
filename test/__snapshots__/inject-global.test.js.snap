// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`injectGlobal composition 1`] = `
".css-1b6ldud {
  background-color: hotpink;
}

html {
  background: pink;
}

html.active {
  color: green;
  background-color: hotpink;
}"
`;

exports[`injectGlobal interpolation 1`] = `
"body {
  color: yellow;
  margin: 0px;
  padding: 0px;
}"
`;

exports[`injectGlobal static 1`] = `
"html {
  background: pink;
}

html.active {
  background: red;
}"
`;
